### PR TITLE
SQ-54/Social screen pt 1

### PR DESCRIPTION
**Pull request by quiro91**
This PR implements the design of the Twitter Feed ([see Zeplin](https://app.zeplin.io/project.html#pid=58a4d9ec7ec20d8738abfa1a&sid=58a8b4e1da81a64280dedd62)). We are dropping the `ListView` and the Twitter SDK Adapter and we're building our own adapter for a brand new `RecyclerView`. 

TODO in the next PR:

- Drop the Twitter date utils and use JodaTime to parse the date
- Highlights mentions and hashtags
- Open the tweet when the tweet is clicked (TBD)
- Load more tweet when the user scrolls down

 Before | After
 --- | ---
before | after

### CODE REVIEW

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Date;
+import java.util.Locale;
+
+import net.squanchy.R;
+import net.squanchy.support.lang.Optional;
+
+public class TwitterDateUtils {
+
+    // Sat Mar 14 02:34:20 +0000 2009
+    private static final SimpleDateFormat DATE_TIME_RFC822 = new SimpleDateFormat("EEE MMM dd HH:mm:ss Z yyyy", Locale.ENGLISH);
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java]: Use `Locale.US` instead just in case, that's what APIs usually use

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    private TwitterDateUtils() {
+    }
+
+    public static String getTimestampFrom(Tweet tweet, Context context) {
+        return getOptionalTimestamp(tweet.createdAt, context)
+                .or(context.getString(R.string.tweet_date_not_available));
+    }
+
+    private static Optional<String> getOptionalTimestamp(String date, Context context) {
+        final String formattedTimestamp;
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java]: Why final? Also you can simply inline the check at the beginning and simplify the code:

```java
if (date == null || !TwitterDateUtils.isValidTimestamp(date)) {
    return Optional.absent();
}

Long createdAtTimestamp = TwitterDateUtils.apiTimeToLong(date);
String formattedTimestamp = TwitterDateUtils.getRelativeTimeString(
    context.getResources(),
    System.currentTimeMillis(),
    createdAtTimestamp
);
return formattedTimestamp;
```

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+     * This method is not thread safe. It has been modified from the original to not rely on global
+     * time state. If a timestamp is in the future we return it as an absolute date string. Within
+     * the same second we return 0s
+     *
+     * @param res               resource
+     * @param currentTimeMillis timestamp for offset
+     * @param timestamp         timestamp
+     * @return the relative time string
+     */
+    private static String getRelativeTimeString(Resources res, long currentTimeMillis, long timestamp) {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java]: The warning here makes me shiver ❌ 

Since we don't even need to have a relative time span (examples are all "[today] at [11:42]" so let's just assume we can do away with a simple format:
 1. `today` or `yesterday` or a date (such as `3rd May 2017`) if older than that
 2. ` on `
 3. the time, such as `11:42`

Besides, this seems to be basically doing what `DateUtils` do already. Use `net.danlew.android.joda.DateUtils` if possible, instead of copying this from somewhere else (Twitter SDK?).

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Locale;
+
+import net.squanchy.R;
+import net.squanchy.support.lang.Optional;
+
+public class TwitterDateUtils {
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java]: I'm not sure you really need this class TBH, seems weird (also please avoid util classes if at all possible, it's a breach of OOP)

File: app/src/main/res/layout/item_tweet.xml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <TextView
+      android:id="@+id/tweet_text"
+      style="@style/Tweet.Card.Title"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:bufferType="spannable"
+      tools:text="Ecco il mio tweet" />
```
Comment (rock3r) [inline app/src/main/res/layout/item_tweet.xml]: English plz

File: app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -23,4 +23,7 @@
   <string name="speaker_list_title">Speakers</string>
   <string name="talks_list_title">Talks</string>
 
+  <string name="droidcon_hashtag" translatable="false">Droidconit</string>
```
Comment (rock3r) [inline app/src/main/res/values/strings.xml]: ❌  why this change? The value comes from the configuration in `team-props`; please don't change things that are working correctly

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
 
-  <string name="activity_social_feed">Social Feed</string>
+  <string name="activity_social_feed">#Droidconit</string>
```
Comment (rock3r) [inline app/src/main/res/values/strings.xml]: What about referencing the string directly?

```xml
<string name="activity_social_feed">@string/social_query</string>
```

**Review by rock3r — CHANGES_REQUESTED** [review]

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java (hunk 4 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java]: Yes, I know, I don't like this code too. This was basically just a copy and paste of `com.twitter.sdk.android.tweetui.TweetDateUtils`. I was already planning of getting read of it as I wrote in the TODO of this PR but if you think it's a blocker I can refactor this code ASAP

Comment (rock3r) [reply app/src/main/java/net/squanchy/tweets/util/TwitterDateUtils.java]: I'd rather not risk this getting anywhere nearer to `develop` than this, if you don't mind doing it now, I'd prefer that

File: app/src/main/java/net/squanchy/tweets/util/TwitterDateFormatter.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+public class TwitterDateFormatter {
+
+    // Sat Mar 14 02:34:20 +0000 2009
+    private static final String DATE_PATTERN = "EEE MMM dd HH:mm:ss Z yyyy";
+
+    private TwitterDateFormatter() {
+    }
+
+    public static String getTimestampFrom(Tweet tweet, Context context) {
+
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/tweets/util/TwitterDateFormatter.java:25]: Why these empty lines? Same in other methods

**Review by rock3r — APPROVED** [review]
LGTM 👍
